<a href="https://colab.research.google.com/github/RanielSosa/predicciones-de-ventas/blob/main/Predicciones_de_ventas.ipynb" target="_parent"><img src="https://colab.research.google.com/assets/colab-badge.svg" alt="Open In Colab"/></a>

In [ ]:
from google.colab import drive
drive.mount('/content/drive')

Drive already mounted at /content/drive; to attempt to forcibly remount, call drive.mount("/content/drive", force_remount=True).


In [184]:
import pandas as pd
filename = '/content/drive/MyDrive/Colab Notebooks/Docs/sales_predictions.csv'
df_sales = pd.read_csv(filename)
df_sales.head()

Item_Identifier  Item_Weight Item_Fat_Content  Item_Visibility  \
0           FDA15         9.30          Low Fat         0.016047   
1           DRC01         5.92          Regular         0.019278   
2           FDN15        17.50          Low Fat         0.016760   
3           FDX07        19.20          Regular         0.000000   
4           NCD19         8.93          Low Fat         0.000000   

               Item_Type  Item_MRP Outlet_Identifier  \
0                  Dairy  249.8092            OUT049   
1            Soft Drinks   48.2692            OUT018   
2                   Meat  141.6180            OUT049   
3  Fruits and Vegetables  182.0950            OUT010   
4              Household   53.8614            OUT013   

   Outlet_Establishment_Year Outlet_Size Outlet_Location_Type  \
0                       1999      Medium               Tier 1   
1                       2009      Medium               Tier 3   
2                       1999      Medium               Tier 1   
3                       1998         NaN               Tier 3   
4                       1987        High               Tier 3   

         Outlet_Type  Item_Outlet_Sales  
0  Supermarket Type1          3735.1380  
1  Supermarket Type2           443.4228  
2  Supermarket Type1          2097.2700  
3      Grocery Store           732.3800  
4  Supermarket Type1           994.7052

In [92]:
df_sales.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 8523 entries, 0 to 8522
Data columns (total 12 columns):
 #   Column                     Non-Null Count  Dtype  
---  ------                     --------------  -----  
 0   Item_Identifier            8523 non-null   object 
 1   Item_Weight                7060 non-null   float64
 2   Item_Fat_Content           8523 non-null   object 
 3   Item_Visibility            8523 non-null   float64
 4   Item_Type                  8523 non-null   object 
 5   Item_MRP                   8523 non-null   float64
 6   Outlet_Identifier          8523 non-null   object 
 7   Outlet_Establishment_Year  8523 non-null   int64  
 8   Outlet_Size                6113 non-null   object 
 9   Outlet_Location_Type       8523 non-null   object 
 10  Outlet_Type                8523 non-null   object 
 11  Item_Outlet_Sales          8523 non-null   float64
dtypes: float64(4), int64(1), object(7)
memory usage: 799.2+ KB


In [185]:
df_sales['Item_Fat_Content']=df_sales['Item_Fat_Content'].replace('LF', 'Low Fat')
df_sales['Item_Fat_Content']=df_sales['Item_Fat_Content'].replace('low fat', 'Low Fat')
df_sales['Item_Fat_Content']=df_sales['Item_Fat_Content'].replace('reg', 'Regular')

In [100]:
df_sales.nunique()

Item_Identifier              1559
Item_Weight                   415
Item_Fat_Content                2
Item_Visibility              7880
Item_Type                      16
Item_MRP                     5938
Outlet_Identifier              10
Outlet_Establishment_Year       9
Outlet_Size                     3
Outlet_Location_Type            3
Outlet_Type                     4
Item_Outlet_Sales            3493
dtype: int64

1) ¿Cuántas filas y columnas?
  8523 filas y 12 columnas.

##2) ¿Cuáles son los tipos de datos en cada variable?
tipos: object, float64 and int64

In [ ]:
len(df_sales)

8523

##3) ¿Hay duplicados? Si es el caso, eliminen algunos duplicados.

No existen duplicados

In [ ]:
df_sales.duplicated().sum()

0

##Ejercicio 4

In [129]:
nulos_Weight=df_sales['Item_Weight'].isnull()
nulos_Size=df_sales['Outlet_Size'].isnull()
df_sales.isnull().sum()

Item_Identifier                 0
Item_Weight                     4
Item_Fat_Content                0
Item_Visibility                 0
Item_Type                       0
Item_MRP                        0
Outlet_Identifier               0
Outlet_Establishment_Year       0
Outlet_Size                  2410
Outlet_Location_Type            0
Outlet_Type                     0
Item_Outlet_Sales               0
dtype: int64

4) Identifiquen los valores faltantes.
  Item_Weight y Outlet_Size son las unicas columnas con valores faltantes.

In [95]:
df_sales.loc[nulos_Weight,'Item_Weight']

7      NaN
18     NaN
21     NaN
23     NaN
29     NaN
        ..
8485   NaN
8487   NaN
8488   NaN
8490   NaN
8504   NaN
Name: Item_Weight, Length: 1463, dtype: float64

##Punto 5

In [ ]:
df_sales.head()

Item_Identifier  Item_Weight Item_Fat_Content  Item_Visibility  \
0           FDA15         9.30          Low Fat         0.016047   
1           DRC01         5.92          Regular         0.019278   
2           FDN15        17.50          Low Fat         0.016760   
3           FDX07        19.20          Regular         0.000000   
4           NCD19         8.93          Low Fat         0.000000   

               Item_Type  Item_MRP Outlet_Identifier  \
0                  Dairy  249.8092            OUT049   
1            Soft Drinks   48.2692            OUT018   
2                   Meat  141.6180            OUT049   
3  Fruits and Vegetables  182.0950            OUT010   
4              Household   53.8614            OUT013   

   Outlet_Establishment_Year Outlet_Size Outlet_Location_Type  \
0                       1999      Medium               Tier 1   
1                       2009      Medium               Tier 3   
2                       1999      Medium               Tier 1   
3                       1998         NaN               Tier 3   
4                       1987        High               Tier 3   

         Outlet_Type  Item_Outlet_Sales  
0  Supermarket Type1          3735.1380  
1  Supermarket Type2           443.4228  
2  Supermarket Type1          2097.2700  
3      Grocery Store           732.3800  
4  Supermarket Type1           994.7052

In [130]:
nulos_Weight

0       False
1       False
2       False
3       False
4       False
        ...  
8518    False
8519    False
8520    False
8521    False
8522    False
Name: Item_Weight, Length: 8523, dtype: bool

### ITEM WEIGHT

###Principal

estoy intentando encontrar el patron de que relacione los articulos con falta de informacion con los articulos que si tienen toda la informacion. hasta ahora, ya se que a los que les falta informacion de una columna no le falta de la otra columna.

In [124]:
Item_Weight_NA=df_sales.loc[nulos_Weight,:]
Item_Weight_NA.isna().sum()


Item_Identifier              0
Item_Weight                  4
Item_Fat_Content             0
Item_Visibility              0
Item_Type                    0
Item_MRP                     0
Outlet_Identifier            0
Outlet_Establishment_Year    0
Outlet_Size                  0
Outlet_Location_Type         0
Outlet_Type                  0
Item_Outlet_Sales            0
dtype: int64

In [186]:
def Upgrade_Item_Identifier_mean(Item_Identifier):
  Item_Identifier_filter=df_sales['Item_Identifier']==Item_Identifier
  values_mean=df_sales[Item_Identifier_filter]['Item_Weight'].mean()
  return values_mean
#Item_Weight_filter_isna =df_sales['Item_Weight'].isna()
df=df_sales
df.loc[nulos_Weight,'Item_Weight']=df_sales.apply(lambda x:Upgrade_Item_Identifier_mean(x['Item_Identifier']),axis=1)
df[nulos_Weight]

Item_Identifier  Item_Weight Item_Fat_Content  Item_Visibility  \
927            FDN52          NaN          Regular         0.130933   
1922           FDK57          NaN          Low Fat         0.079904   
4187           FDE52          NaN          Regular         0.029742   
5022           FDQ60          NaN          Regular         0.191501   

         Item_Type  Item_MRP Outlet_Identifier  Outlet_Establishment_Year  \
927   Frozen Foods   86.9198            OUT027                       1985   
1922   Snack Foods  120.0440            OUT027                       1985   
4187         Dairy   88.9514            OUT027                       1985   
5022  Baking Goods  121.2098            OUT019                       1985   

     Outlet_Size Outlet_Location_Type        Outlet_Type  Item_Outlet_Sales  
927       Medium               Tier 3  Supermarket Type3          1569.9564  
1922      Medium               Tier 3  Supermarket Type3          4434.2280  
4187      Medium               Tier 3  Supermarket Type3          3453.5046  
5022       Small               Tier 1      Grocery Store           120.5098

###Prueba

In [177]:
Item_Identifier_filter=df_sales['Item_Identifier']=='FDO28'
Item_Type_filter=df_sales['Item_Type']=='Frozen Foods'
Outlet_Size_filter=df_sales['Outlet_Size']=='Supermarket Type3'
Outlet_Location_Type_filter=df_sales['Outlet_Location_Type']=='Tier 3'
Item_MRP_filter=df_sales['Item_MRP']==121.6098
test=df_sales.loc[Item_Identifier_filter,:]
test

Item_Identifier  Item_Weight Item_Fat_Content  Item_Visibility  \
465            FDO28        5.765          Low Fat         0.072285   
5485           FDO28          NaN          Low Fat         0.126585   
5549           FDO28        5.765          Low Fat         0.072238   
8185           FDO28        5.765          Low Fat         0.072411   
8444           FDO28          NaN          Low Fat         0.071948   

         Item_Type  Item_MRP Outlet_Identifier  Outlet_Establishment_Year  \
465   Frozen Foods  119.9098            OUT035                       2004   
5485  Frozen Foods  122.4098            OUT019                       1985   
5549  Frozen Foods  120.0098            OUT013                       1987   
8185  Frozen Foods  122.5098            OUT049                       1999   
8444  Frozen Foods  121.6098            OUT027                       1985   

     Outlet_Size Outlet_Location_Type        Outlet_Type  Item_Outlet_Sales  
465        Small               Tier 2  Supermarket Type1          2048.6666  
5485       Small               Tier 1      Grocery Store           120.5098  
5549        High               Tier 3  Supermarket Type1          1687.1372  
8185      Medium               Tier 1  Supermarket Type1          1084.5882  
8444      Medium               Tier 3  Supermarket Type3           482.0392

In [131]:
df_sales.loc[nulos_Weight,'Item_Weight']

927    NaN
1922   NaN
4187   NaN
5022   NaN
Name: Item_Weight, dtype: float64

In [141]:
Item_Weight_NA['Item_Type'].value_counts()

Fruits and Vegetables    213
Snack Foods              212
Household                151
Frozen Foods             138
Dairy                    116
Baking Goods             112
Canned                   110
Health and Hygiene        90
Meat                      88
Soft Drinks               71
Breads                    47
Others                    32
Hard Drinks               31
Breakfast                 21
Starchy Foods             18
Seafood                   13
Name: Item_Type, dtype: int64

In [132]:
Item_Weight_NA.nunique()

Item_Identifier              1142
Item_Weight                   409
Item_Fat_Content                2
Item_Visibility              1362
Item_Type                      16
Item_MRP                     1371
Outlet_Identifier               2
Outlet_Establishment_Year       1
Outlet_Size                     2
Outlet_Location_Type            2
Outlet_Type                     2
Item_Outlet_Sales            1183
dtype: int64

In [ ]:
Item_Weight_NA['Item_Fat_Content']=Item_Weight_NA['Item_Fat_Content'].replace('LF', 'Low Fat')
Item_Weight_NA['Item_Fat_Content']=Item_Weight_NA['Item_Fat_Content'].replace('low fat', 'Low Fat')
Item_Weight_NA['Item_Fat_Content']=Item_Weight_NA['Item_Fat_Content'].replace('reg', 'Regular')

In [133]:
#Item_Weight_NA.groupby(['Outlet_Type','Outlet_Location_Type','Item_Type']).count()
Item_Weight_NA.groupby(['Outlet_Type','Outlet_Location_Type','Item_Fat_Content','Outlet_Identifier','Outlet_Size','Item_Identifier']).count()

Item_Weight  \
Outlet_Type       Outlet_Location_Type Item_Fat_Content Outlet_Identifier Outlet_Size Item_Identifier                
Grocery Store     Tier 1               Low Fat          OUT019            Small       DRC25                      1   
                                                                                      DRD15                      1   
                                                                                      DRD25                      1   
                                                                                      DRD27                      1   
                                                                                      DRD60                      1   
...                                                                                                            ...   
Supermarket Type3 Tier 3               Regular          OUT027            Medium      FDZ47                      1   
                                                                                      FDZ49                      1   
                                                                                      FDZ50                      1   
                                                                                      FDZ51                      1   
                                                                                      FDZ59                      1   

                                                                                                       Item_Visibility  \
Outlet_Type       Outlet_Location_Type Item_Fat_Content Outlet_Identifier Outlet_Size Item_Identifier                    
Grocery Store     Tier 1               Low Fat          OUT019            Small       DRC25                          1   
                                                                                      DRD15                          1   
                                                                                      DRD25                          1   
                                                                                      DRD27                          1   
                                                                                      DRD60                          1   
...                                                                                                                ...   
Supermarket Type3 Tier 3               Regular          OUT027            Medium      FDZ47                          1   
                                                                                      FDZ49                          1   
                                                                                      FDZ50                          1   
                                                                                      FDZ51                          1   
                                                                                      FDZ59                          1   

                                                                                                       Item_Type  \
Outlet_Type       Outlet_Location_Type Item_Fat_Content Outlet_Identifier Outlet_Size Item_Identifier              
Grocery Store     Tier 1               Low Fat          OUT019            Small       DRC25                    1   
                                                                                      DRD15                    1   
                                                                                      DRD25                    1   
                                                                                      DRD27                    1   
                                                                                      DRD60                    1   
...                                                                                                          ...   
Supermarket Type3 Tier 3               Regular       

### OUTLET SIZE

In [183]:
Outlet_Size_NA=df_sales.loc[nulos_Size,:]
Outlet_Size_NA.isna().sum()

Item_Identifier                 0
Item_Weight                  2410
Item_Fat_Content                0
Item_Visibility                 0
Item_Type                       0
Item_MRP                        0
Outlet_Identifier               0
Outlet_Establishment_Year       0
Outlet_Size                  2410
Outlet_Location_Type            0
Outlet_Type                     0
Item_Outlet_Sales               0
dtype: int64

In [182]:
def Upgrade_Item_Identifier_mean(Item_Identifier):
  Item_Identifier_filter=df_sales['Item_Identifier']==Item_Identifier
  values_mean=df_sales[Item_Identifier_filter]['Outlet_Size']
  return values_mean
#Item_Weight_filter_isna =df_sales['Item_Weight'].isna()
df=df_sales
df.loc[nulos_Size,'Item_Weight']=df_sales.apply(lambda x:Upgrade_Item_Identifier_mean(x['Item_Identifier']),axis=1)
df[nulos_Size]

Item_Identifier  Item_Weight Item_Fat_Content  Item_Visibility  \
3              FDX07          NaN          Regular         0.000000   
8              FDH17          NaN          Regular         0.016687   
9              FDU28          NaN          Regular         0.094450   
25             NCD06          NaN          Low Fat         0.099887   
28             FDE51          NaN          Regular         0.161467   
...              ...          ...              ...              ...   
8502           NCH43          NaN          Low Fat         0.070712   
8508           FDW31          NaN          Regular         0.043246   
8509           FDG45          NaN          Low Fat         0.214306   
8514           FDA01          NaN          Regular         0.054489   
8519           FDS36          NaN          Regular         0.046982   

                  Item_Type  Item_MRP Outlet_Identifier  \
3     Fruits and Vegetables  182.0950            OUT010   
8              Frozen Foods   96.9726            OUT045   
9              Frozen Foods  187.8214            OUT017   
25                Household   45.9060            OUT017   
28                    Dairy   45.5086            OUT010   
...                     ...       ...               ...   
8502              Household  216.4192            OUT045   
8508  Fruits and Vegetables  199.4742            OUT045   
8509  Fruits and Vegetables  213.9902            OUT010   
8514                 Canned   57.5904            OUT045   
8519           Baking Goods  108.1570            OUT045   

      Outlet_Establishment_Year Outlet_Size Outlet_Location_Type  \
3                          1998         NaN               Tier 3   
8                          2002         NaN               Tier 2   
9                          2007         NaN               Tier 2   
25                         2007         NaN               Tier 2   
28                         1998         NaN               Tier 3   
...                         ...         ...                  ...   
8502                       2002         NaN               Tier 2   
8508                       2002         NaN               Tier 2   
8509                       1998         NaN               Tier 3   
8514                       2002         NaN               Tier 2   
8519                       2002         NaN               Tier 2   

            Outlet_Type  Item_Outlet_Sales  
3         Grocery Store           732.3800  
8     Supermarket Type1          1076.5986  
9     Supermarket Type1          4710.5350  
25    Supermarket Type1           838.9080  
28        Grocery Store           178.4344  
...                 ...                ...  
8502  Supermarket Type1          3020.0688  
8508  Supermarket Type1          2587.9646  
8509      Grocery Store           424.7804  
8514  Supermarket Type1           468.7232  
8519  Supermarket Type1           549.2850  

[2410 rows x 12 columns]

### Continuacion

In [ ]:
df_sales.nunique()

Item_Identifier              1559
Item_Weight                   415
Item_Fat_Content                5
Item_Visibility              7880
Item_Type                      16
Item_MRP                     5938
Outlet_Identifier              10
Outlet_Establishment_Year       9
Outlet_Size                     3
Outlet_Location_Type            3
Outlet_Type                     4
Item_Outlet_Sales            3493
dtype: int64

In [ ]:
df_sales[['Item_Weight','Outlet_Size']]

Item_Weight Outlet_Size
0           9.300      Medium
1           5.920      Medium
2          17.500      Medium
3          19.200         NaN
4           8.930        High
...           ...         ...
8518        6.865        High
8519        8.380         NaN
8520       10.600       Small
8521        7.210      Medium
8522       14.800       Small

[8523 rows x 2 columns]

In [ ]:
valores=df_sales.groupby(['Item_Type','Outlet_Type'])[['Item_Identifier','Item_Weight','Outlet_Size']].count()
valores

Item_Identifier  Item_Weight  Outlet_Size
Item_Type     Outlet_Type                                                 
Baking Goods  Grocery Store                   85           42           43
              Supermarket Type1              426          426          283
              Supermarket Type2               68           68           68
              Supermarket Type3               69            0           69
Breads        Grocery Store                   33           17           16
...                                          ...          ...          ...
Soft Drinks   Supermarket Type3               45            0           45
Starchy Foods Grocery Store                   13            9            4
              Supermarket Type1              104          104           70
              Supermarket Type2               17           17           17
              Supermarket Type3               14            0           14

[64 rows x 3 columns]

5) Decidan cómo abordar los valores faltantes y cómo hacerlo. (Esto requiere sus criterios, así que expliquen su elección).


Mi propuesta es agrupar los registros por tipo de producto y otras variables que me ayuden a identificar un patron en el producto no identificado y una vez encontrado el patron realizar la imputacion de los datos usando como referencia los valores similares a los articulos con valores faltantes, pero aun no he podido encontrar la forma correcta de relacionar los articulos en grupos que me puedan dar informacion muy especifica del conjunto de datos flatantes.

estoy usando el metodo 'Nunique' para identificar cuales columnas tienen datos utilizables pero con pocos valores unicos.

Tambien estoy agrupando por el set de datos con menos valores unicos para entender que informacion obtengo y luego tomar varios registros al azar de pueba para validar que la relacion es la mas eficiente para imputar y en caso de que no exista nada que relacione los registros con datos faltantes con los registros con todos sus datos entonces esos datos hay que eliminarlos.

##6) Confirmen que no hay valores faltantes después de abordarlos.

Si el planteamiento es correcto, no existen datos faltantes, aunque pueda que existan unos cuantos registros menos que la cantidad inicial

7) Encuentren y arreglen alguna categoría inconsistente de datos (example: fix cat, Cat, and cats so that they are consistent) 

In [ ]:
df_sales.nunique()

Item_Identifier              1559
Item_Weight                   415
Item_Fat_Content                5
Item_Visibility              7880
Item_Type                      16
Item_MRP                     5938
Outlet_Identifier              10
Outlet_Establishment_Year       9
Outlet_Size                     3
Outlet_Location_Type            3
Outlet_Type                     4
Item_Outlet_Sales            3493
dtype: int64

para determinar cual categoria contenia valores mal escritos, volvi a ver las columnas con valores unicos y luego fuí probando con las que tenian valores unicos por debajo de 10

In [ ]:
df_sales['Item_Fat_Content'].value_counts()

Low Fat    5517
Regular    3006
Name: Item_Fat_Content, dtype: int64

In [ ]:
df_sales['Item_Fat_Content']=df_sales['Item_Fat_Content'].replace('LF', 'Low Fat')
df_sales['Item_Fat_Content']=df_sales['Item_Fat_Content'].replace('low fat', 'Low Fat')
df_sales['Item_Fat_Content']=df_sales['Item_Fat_Content'].replace('reg', 'Regular')

In [ ]:
df_sales.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 8523 entries, 0 to 8522
Data columns (total 12 columns):
 #   Column                     Non-Null Count  Dtype  
---  ------                     --------------  -----  
 0   Item_Identifier            8523 non-null   object 
 1   Item_Weight                7060 non-null   float64
 2   Item_Fat_Content           8523 non-null   object 
 3   Item_Visibility            8523 non-null   float64
 4   Item_Type                  8523 non-null   object 
 5   Item_MRP                   8523 non-null   float64
 6   Outlet_Identifier          8523 non-null   object 
 7   Outlet_Establishment_Year  8523 non-null   int64  
 8   Outlet_Size                6113 non-null   object 
 9   Outlet_Location_Type       8523 non-null   object 
 10  Outlet_Type                8523 non-null   object 
 11  Item_Outlet_Sales          8523 non-null   float64
dtypes: float64(4), int64(1), object(7)
memory usage: 799.2+ KB


8) Para cualquier columna numérica, obtengan las estadísticas resumidas para cada uno (mínimo, máximo y media)


In [ ]:
df_sales['Outlet_Establishment_Year'].mean()

1997.8318667135984

In [ ]:
df_sales['Outlet_Establishment_Year'].min()

1985

In [ ]:
df_sales['Outlet_Establishment_Year'].max()

2009